docs: Unify devmapper option docs in docs/man/docker.1.md

cgwalters opened the pull request:
I'll say up front, `man docker` is probably not the best place to
include all of the details of the devmapper storage options.  However,
today we have them there, and _also_ in the `README.md` in the source
code.

There was quite a lot of duplication.  I went over the two docs and
took the best from each.

The man page makes sense because it's shipped with the package/system,
and it's important for system administrators to be able to see options
documented for the version of the system that they have, rather than
just looking on github or whatever.

As a future step I'd like to move the devmapper options out to
e.g. `man docker-devmapper` or so, but let's first deduplicate.

Docker-DCO-1.1-Signed-off-by: Colin Walters walters@verbum.org (github: cgwalters)

cgwalters committed:
docs: Unify devmapper option docs in docs/man/docker.1.md

I'll say up front, `man docker` is probably not the best place to
include all of the details of the devmapper storage options.  However,
today we have them there, and also in the `README.md` in the source
code.

Finally, we *also* have them in the CLI docs, which this patch doesn't
address.

There was quite a lot of duplication between the README.md and the man
page.  I went over the two docs and took the best from each.

The man page makes sense because it's shipped with the package/system,
and it's important for system administrators to be able to see options
documented for the version of the system that they have, rather than
just looking on github or whatever.

As a future step I'd like to move the devmapper options out to
e.g. `man docker-devmapper` or so, and to unify that with the CLI
docs, but let's first deduplicate.

Docker-DCO-1.1-Signed-off-by: Colin Walters <walters@verbum.org> (github: cgwalters)

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "devmapper-docs-merge-to-manpage" git@github.com:cgwalters/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

vbatts commented:
i'm not terribly opposed, though these options are also duplicated in the [cli docs](https://github.com/docker/docker/blob/master/docs/sources/reference/commandline/cli.md)

rhatdan commented:
I think we should at least get this into docker server side.

cgwalters commented:
Aaaah...there's *three* places these docs live?  Okay, so who should win between the man page and the cli docs?

thaJeztah commented:
Tough one, I'm personally for *both*; I think there's really two camps here; people that prefer man pages for reference, and those that like to have the docs open while working with Docker.

In addition, Google does a terrible job at indexing the man-pages for those looking for answers :smiley:

Also, there's a proposal for adding links to the docs in docker help output; https://github.com/docker/docker/issues/13298

/cc @bfirsh thoughts?

thaJeztah commented:
Having said that, I absolutely agree that we should try to reduce duplicated docs in the online documentation; especially between the CLI docs and Docker Run reference. Perhaps even a dedicated page for each command, and convert the CLI docs to a basic "index" linking to those pages.

SvenDowideit commented:
The `cli.md` docs should always win - if you're insisting on choosing one.

@moxiegirl is working on replacing the current mkdocs tooling with hugo, so that may help de-duplicate in future, but I'd suggest the information really does need to be in both.

thaJeztah commented:
> wrt further re-organising the docs into separate pages etc - I'd defer that to the DockerCon get together, when we can all agree on timeframes.

Sounds good. You're sending me tickets? 😉

All jokes aside, fully agree that's a discussion for the longer term, and no direct issue.

cgwalters commented:
Ok, so given current consensus is to keep both the man page and the cli, can we move forward with killing the devmapper `README.md` copy?

thaJeztah commented:
@vbatts I think it's your call on removing it from the README.md

vbatts commented:
It will definitely make them more accessible. Let's add links to the online
docs and referral to the docker(1) man page from the readme
On May 20, 2015 11:58 AM, "Sebastiaan van Stijn" <notifications@github.com>
wrote:

> @vbatts <https://github.com/vbatts> I think it's your call on removing it
> from the README.md
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13323#issuecomment-103937826>.
>

cgwalters commented:
I tweaked the `README.md` and the commit message to link to both the CLI docs and the man page so people know to edit both, until we figure out how to dedup those.

thaJeztah commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+requires no pre-configuration, but is extremely inefficient.  Do not
+use it in production.
+
+In oder to make the best use of Docker with the devicemapper backend,
```
s/oder/order/

thaJeztah replied:
^^ not really reviewing yet, just caught my eye

thaJeztah commented:
>I tweaked the README.md and the commit message to link to both the CLI docs and the man page so people know to edit both, until we figure out how to dedup those.

Good thinking. Thanks!

Are you planning on updating the `cli.md` docs as part of this PR, or only the changes that we have now?

In other words; let us know when it's ready for review.

cgwalters committed:
docs/cli.md: Sync devicemapper options with docs/docker.1.md

Currently, these two files both document the command line options
available.  There is some discussion of deduplicating, but no solid
plan yet.

In the mean time, this patch synchronizes some of the most important
changes in the man page into the CLI docs.  In particular, we want to
document and emphasize `dm.thinpooldev`, and note that loopback is
deprecated.

Docker-DCO-1.1-Signed-off-by: Colin Walters <walters@verbum.org> (github: cgwalters)

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "devmapper-docs-merge-to-manpage" git@github.com:cgwalters/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

cgwalters commented:
Fixed the typo `oder`, and added a secondary patch which syncs up the `cli.md`.  I think this is ready to review.

thaJeztah commented:
Thanks! I'll try and give this a review soon.

ping @fredlf @jamtur01 @moxiegirl @SvenDowideit this PR is now ready for review

moxiegirl commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+Specifies a custom blockdevice to use for the thin pool.
+
+If using a block device for device mapper storage, ideally lvm2 would
+be used to create/manage the thin-pool volume that is then handed to
```
This sentence is awkward.  Split into two sentences to reduce complexity and increase clarity. Maybe:

If using a block device for device mapper storage, it is best to use `lvm2` to create/manage the thin-pool volume. This volume is then handed to
Docker to exclusively create/manage the thin and thin snapshot volumes
needed for its containers.

moxiegirl commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
-options are "ext4" and "xfs". The default is "ext4"
+
+Specifies the filesystem type to use for the base device. The
+supported options are "ext4" and "xfs". The default is "ext4"
```
Use ticks here please.

Commit:
docs: Fix typos and clarity/wording in devicemapper manpage and cli

(And some further tweaks by Colin Walters <walters@verbum.org>)

Signed-off-by: Mary Anthony <mary@docker.com>

moxiegirl commented:
Nice contribution @cgwalters thank you for doing the work. I found some typos spelling and missing word errors.  Instead of detailing them in comments (I did some) I created a patch:

https://gist.github.com/moxiegirl/81f988d6294c445d282d

Save the gists content to a patch file and apply it. I used this http://makandracards.com/makandra/2521-git-how-to-create-and-apply-patches as a reference.

cgwalters commented:
Hi @moxiegirl, thanks for the patch!  I'm not entirely sure of the expected procedure here, but I applied your patch locally to my PR, added a few tweaks on top, changed your commit message to be a bit more informative, and pushed an update.

I believe you addressed all comments you had with my original patch, and hopefully you agree with my edits to yours on top, so let me know if there's anything else I need to do.

Thanks!

moxiegirl commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -316,22 +333,28 @@ Currently supported options of `devicemapper`:
-    Specifies the size to use when creating the loopback file for the "data"
-    device which is used for the thin pool. The default size is 100G. Note that
-    the file is sparse, so it will not initially take up this much space.
+		>**Note**: This option configures devicemapper loopback, which is not for
```
Looks like an issue here in the Markdow --- it might have come from me. Can you fix it.

![image](https://cloud.githubusercontent.com/assets/1347209/7820082/e33e34ca-039a-11e5-801c-77c1d091a0fc.png)

moxiegirl commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -316,22 +333,28 @@ Currently supported options of `devicemapper`:
-    "metadata" device which is used for the thin pool. The default size is 2G.
-    Note that the file is sparse, so it will not initially take up this much
-    space.
+		>**Note**: This option configures devicemapper loopback, which is not for
```
same problem here ...too far indented it is being treated as code

moxiegirl commented:
With fixes to the note Markdown, LGTM

cgwalters committed:
cli.md: Use consistent indents for devmapper options and use ``

We had some corrupted indentation from patch back and forth; change
all entries in the devmapper table to use 4 spaces which appears
to be the standard.  Or at least as standard as markdown gets ;)

Also use the code quoter "``" more consistently.

Docker-DCO-1.1-Signed-off-by: Colin Walters <walters@verbum.org> (github: cgwalters)

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "devmapper-docs-merge-to-manpage" git@github.com:cgwalters/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

cgwalters commented:
Looks like we had a lot of tab/space confusion and other issues.  I took a stab and fixing things, although there are some pre-existing issues like the text with `Udev Sync Supported: true` not being a code block that I didn't fix in this.  I'm trying to incrementally improve the current situation without going too far down the rabbit hole of rewriting the docs =)

vbatts commented:
Rewrite all the things!

;-)
On May 26, 2015 21:45, "Colin Walters" <notifications@github.com> wrote:

> Looks like we had a lot of tab/space confusion and other issues. I took a
> stab and fixing things, although there are some pre-existing issues like
> the text with Udev Sync Supported: true not being a code block that I
> didn't fix in this. I'm trying to incrementally improve the current
> situation without going too far down the rabbit hole of rewriting the docs
> =)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13323#issuecomment-105716764>.
>

moxiegirl commented:
Thanks @cgwalters, that looks sorted.  We require two LGTMs for documentation. ping @thaJeztah

thaJeztah commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+use it in production.
+
+To make the best use of Docker with the devicemapper backend,
+you must have a recent version of `lvm2`.  Use `lvm` to create a thin
```
"Use `lvm` to create.."; I guess this should be "Use `lvm2` to create...", or is the CLI utility always called `lvm`?

cgwalters replied:
The CLI is always called `lvm`, `lvm2` just happens to be the package name at least on Fedora.  It's probably saner to just say `lvm` everywhere and people can figure out the package name.  Will fix.

thaJeztah commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+changing thin-pool features, automatic thinp metadata checking when lvm2 activates
+the thin-pool, etc.
+
+Example use: ``docker -d --storage-opt dm.thinpooldev=/dev/mapper/thin-pool``
```
Wondering why you're using double back-ticks here (and some other places below)? 

It could well be valid Markdown, but I'm not familiar with it. GitHub seems to render this correctly, but have you checked if the man-page is generated correctly?

thaJeztah commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -302,80 +319,74 @@ Currently supported options of `devicemapper`:
-    Example use:
-
-        $ docker -d --storage-opt dm.basesize=20G
+    Example use: ``$ docker -d --storage-opt dm.basesize=20G``
```
This changes the example from a 'code block' to an inline example (here, and below).

I understand that (for maintenance), it's easier to have this the same as the man-pages, but it's inconsistent with the other examples in this document.

I'd prefer them to be consistent here. Could you change the "Example use" lines in this document back to the old formatting?

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
 
-Options to storage backend can be specified with **--storage-opt** flags. The
-only backend which currently takes options is *devicemapper*. Therefore use these
+Docker uses storage backends (or "graphdriver" in Docker internals) to create
```
the (or ... ) seems a little weird. Perhaps:

(known as graph drivers in the Docker internals)

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+the size of images and containers. The default value is 10G. Note,
+thin devices are inherently "sparse", so a 10G device which is mostly
+empty doesn't use 10 GB of space on the pool. However, the filesystem
+will use more space for the empty case the larger the device
```
"empty case" doesn't make sense to me?

jamtur01 replied:
empty filesystem?

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+
+#### dm.use_deferred_removal
+
+Enables use of deferred device removal if `libdm` and kernel driver
```
and the kernel driver

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+
+Deferred device removal means that if device is busy when devices are
+being removed/deactivated, then a deferred removal is scheduled on
+device. And devices automatically goes away when last user of device
```
And devices automatically go away when the last user of the device exits.

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+exits.
+
+For example, when a container exits, its associated thin device is removed. If
+that devices has leaked into some other mount namespace and can't be removed,
```
device

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+For example, when a container exits, its associated thin device is removed. If
+that devices has leaked into some other mount namespace and can't be removed,
+the container exit still succeeds and this option causes the system to schedule
+device for deferred removal. It does not wait in a loop trying to remove a busy
```
to schedule the device

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+
+#### dm.loopdatasize
+
+**Note**: This option configures devicemapper loopback, which is not for
```
which is not recommended for

cgwalters replied:
How about "should not be used in production"?

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+
+Specifies the size to use when creating the loopback file for the
+"data" device which is used for the thin pool. The default size is
+100G. Note that the file is sparse, so it will not initially take up
```
If this is a note then please format as a note.

thaJeztah replied:
@jamtur01 this is the man-page, I think a slightly different formatting is used there.

jamtur01 replied:
@thaJeztah Nope - man pages have notes - see two paragraphs below.

thaJeztah commented:
Thanks, @cgwalters this is looking quite good. I added some comments, mainly around "formatting" on the online Command Line Reference page

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+
+#### dm.loopmetadatasize
+
+**Note**: This option configures devicemapper loopback, which is not for
```
As above.

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+
+Specifies the size to use when creating the loopback file for the
+"metadadata" device which is used for the thin pool. The default size
+is 2G. Note that the file is sparse, so it will not initially take up
```
As above

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
-#### dm.metadatadev
-Specifies a custom blockdevice to use for metadata for the thin pool.
+Specifies a custom blockdevice to use for data for a
+Docker-managed thin pool.  It is better to use `dm.thinpooldev` which
```
Why is it better?

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
-      --storage-opt dm.basesize=20G
+
+Enables or disables the use of `blkdiscard` when removing devicemapper
+devices.  This is enabled by default (only) if using loopback devices
```
(only) in parentheses is odd - perhaps rewrite?

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+#### dm.override_udev_sync_check
+
+Overrides the `udev` synchronization checks between `devicemapper` and
+`udev`.  `udev` is the device manager for the Linux kernel.
```
Perhaps don't start sentence with a command?

jamtur01 commented on docs/man/docker.1.md:
```diff
@@ -292,79 +292,192 @@ inside it)
+
+        $ docker -d --storage-opt dm.override_udev_sync_check=true
+
+When this value is `true`, the `devicemapper` continues and simply
```
the devicemapper? the devicemapper driver?

jamtur01 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -302,80 +319,74 @@ Currently supported options of `devicemapper`:
-    Example use:
+    Specifies the size to use when creating the loopback file for the
+    "metadadata" device which is used for the thin pool. The default size
+    is 2G. Note that the file is sparse, so it will not initially take up
```
As per other note comments.

cgwalters committed:
docs: Address review comments in docker.1.md and cli.md

 - Revert back to indented blocks for cod examples in cli.md for consistency.
 - Drop use of double backquote
 - Reword parts for clarity
 - Fix typos

Docker-DCO-1.1-Signed-off-by: Colin Walters <walters@verbum.org> (github: cgwalters)

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "devmapper-docs-merge-to-manpage" git@github.com:cgwalters/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~5
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

cgwalters commented:
Hi @jamtur01 - The follow up patch above should address all your comments.  Though can I ask that if we're reviewing these docs more extensively, can we do it as a separate PR so I can land the deduplication, then we improve from there?

LK4D4 commented:
@cgwalters Would you mind to rebase?

cgwalters commented:
@LK4D4 Rebased.

cgwalters commented:
If this patch stays out it's going to keep accumulating merge conflicts =(

rhatdan commented:
LGTM  :^)

cgwalters commented:
Maybe it would make sense to just transfer this PR to the docs team?  @moxiegirl  ?

moxiegirl commented:
@cgwalters Given the recent break up of the cli.md, it does make sense for us to carry.  Thanks @cgwalters for doing such terrific ground breaking.
